pkg/hubble/dropeventemitter/dropeventemitter.go

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dropeventemitter
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	metaslimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	"github.com/cilium/cilium/pkg/identity"
+	client "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+type DropEventEmitter struct {
+	history     map[string]time.Time
+	historySize int
+	interval    time.Duration
+	broadcaster record.EventBroadcaster
+	recorder    record.EventRecorder
+}
+
+func NewDropEventEmitter(interval time.Duration, historySize int, k8s client.Clientset) *DropEventEmitter {
+	emitter := DropEventEmitter{
+		history:     map[string]time.Time{},
+		historySize: historySize,
+		interval:    interval,
+		broadcaster: record.NewBroadcaster(),
+	}
+	emitter.broadcaster.StartRecordingToSink(&typedv1.EventSinkImpl{Interface: k8s.CoreV1().Events("")})
+	emitter.recorder = emitter.broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "cilium"})
+
+	return &emitter
+}
+
+func (e *DropEventEmitter) expireHistory() {
+	now := time.Now()
+	for combo, timestamp := range e.history {
+		// Very simple rate limiter. Also remove future entries due to DST.
+		if now.Add(-e.interval).After(timestamp) || now.Before(timestamp) {
```

Comment:
It seems to me that the second part of the condition cannot evaluate to `true` and if correct then I'll rather remove it to avoid any confusion (nit on the way).
```suggestion
		if timestamp.Add(e.interval).Before(now) {
```

Reply:
Yes, it can be, when daylight savings time sets the clock back for an hour.

Reply:
The instants in time represented by `time.Time` values are not timezone sensitive (they store location for presentation purposes only) and use a monotonic clock for comparison and add/subtract operations (see [Monotonic Clocks](https://pkg.go.dev/time#hdr-Monotonic_Clocks)). So I think the second part of the condition isn't necessary.

Reply:
I've replaced `record.NewBroadcaster()` with `record.NewBroadcasterWithCorrelatorOptions()` and options, eliminating the need for my own rate limiter.